tests/python/unittest/test_meta_schedule_vnni_integration.py

```diff
@@ -0,0 +1,249 @@
+            relay.create_executor(
+                "vm",
+                mod=relay_mod,
+                device=dev,
+                target="llvm",
+            )
+            .evaluate()(data, params["weight"], params["bias"])
+            .numpy()
+        )
+        runtime = tvm.contrib.graph_executor.GraphModule(lib["default"](dev))
+        runtime.set_input("data", data)
+        runtime.run()
+        out = runtime.get_output(0).numpy()
+        np.testing.assert_equal(out, ref)
+
+    return relay_mod, params, f_check
+
+
+@pytest.mark.skip("Requires cascadelake")
+def test_vnni_schedule_fn_database():
+    m, n, k = 1024, 1024, 1024
+    target = tvm.target.Target("llvm -mcpu=cascadelake -num-cores 4")
+    dev = tvm.cpu(0)
+    relay_mod, params, f_check = _relay_dense(m, n, k)
+
+    with ms.database.ScheduleFnDatabase(
+        _schedule_dense(
+            m=m,
+            do_tune=False,
+        )
+    ), tvm.transform.PassContext(
+        opt_level=3,
+        config={"relay.backend.use_meta_schedule": True},
+    ):
+        # pylint: disable=W0105
+        """The log should say
+        Warning: Cannot find workload: tvmgen_default_fused_expand_dims
+        Warning: Cannot find workload: tvmgen_default_fused_cast
+        Warning: Cannot find workload: tvmgen_default_fused_cast_1
+        Warning: Cannot find workload: tvmgen_default_fused_nn_batch_matmul
+
+        This means batch matmul and others are scheduled by TE, and dense (the one not warned)
+        is found in the meta schedule tuning database during compilation
+        """
+        # pylint: enable=W0105
+        lib = relay.build(relay_mod, target=target, params=params)
+    f_check(lib, dev)
+
+
+@pytest.mark.skip("Requires cascadelake")
+def test_vnni_schedule_fn_tune():
+    # pylint: disable=W0105
+    """
+    We can inject and apply a custom TIR scheduling to a TE compute of interest, using
+    the "schedule_rule" annotation. For example, in topi/x86/dense.py we have the following
+    declaration for int8 dense targeting the VNNI instruction.
+
+    C = te.compute(
+        ...
+        attrs={"schedule_rule": "meta_schedule.dense_vnni"},
+    )
+
+    When the MetaSchedule encounters a TensorIR block with the "schedule_rule" annotation,
+    it looks up the packed func registry for a function that is associated with the given schedule
+    rule key ("meta_schedule.dense_vnni" in this example). The signature of such custom schedule
+    functions must be
+
+       (tir.schedule.Schedule, tir.schedule.BlockRV) -> [tir.schedule.Schedule].
+
+    The BlockRV argument corresponds to the TE compute annotated with "schedule_rule".
+
+    The relevant code is in meta_schedule/space_generator/post_order_apply.cc.
+    """
+
+    def schedule_rule_dense_vnni(sch: Schedule, dense_block: BlockRV):
+        _schedule_dense(m=None, do_tune=True)(sch, dense_block)
+        return [sch]
+
+    register_func("meta_schedule.dense_vnni", schedule_rule_dense_vnni)
```

Comment:
@junrushao @masahi or others, may I ask what the difference is between using the TE annotation as described (e.g. `attrs={"schedule_rule": "meta_schedule.dense_vnni"},` and a corresponding packed func defining the schedule to use, as opposed to just generating the space via 

```
space=ms.space_generator.ScheduleFn(
     _schedule_dense,
    ...
),
```
?

Is it that in this test case we allow auto scheduling for all ops but apply special manual scheduling for certain ops (dense in this case), whereas if we use the ScheduleFn technique for generating a search space we do not allow other operators to be auto scheduled? Thanks!

Reply:
I think `ScheduleFnDatabase` is for a completely manual schedule, while the `register_func` way allows autotvm-style template based tuning. At least that's what I wanted to demonstrate before this PR or before `ScheduleFnDatabase` was introduced.

Reply:
In this case I'm not referring to ScheduleFnDatabase as is used in `test_vnni_schedule_fn_database`. I'm referring here to what is done in the test `test_vnni_schedule_fn_tune` which utilizes the TE compute `schedule_rule` attr annotation along with a global packed function for the schedule that matches the annotation value `meta_schedule.dense_vnni`. I'm wondering if there is any difference or advantage between using the TE attr annotation and packed func as opposed to specifying an alternate search space with ScheduleFn.

Reply:
Hi Chris, `ScheduleFn` space generator is designed to schedule all blocks in the whole Schedule, not block specific. The annotation based packfunc scheduling only works in `PostOrderApply` space generator, which essentially applies this annotated rule for this specific block, and apply default schedule rules (or schedule rules given in user interface) to other non-specified blocks.

Therefore, creating a `ScheduleFn` takes more effort and use the annotation based scheduling is easier because you don't need to worry about scheduling of other blocks.

Reply:
Ahh, okay I see, thanks for the discussion @zxybazh @masahi, this is helpful.